File: InterviewCakeProblems2/reverse_words.py
def reverse_words(message):
    message = reverse_string_in_place(message, 0, len(message) - 1)

    startIdx = 0
    for i in range(len(message) + 1):
        if i == len(message) or message[i] == " ":
            reverse_string_in_place(message, startIdx, i - 1)
            startIdx = i + 1
    return "".join(message)


def reverse_string_in_place(chars, start, end):

    while start < end:
        chars[start], chars[end] = chars[end], chars[start] 
        start += 1
        end -= 1
    return chars


message = ['c', 'a', 'k', 'e', ' ',
           'p', 'o', 'u', 'n', 'd', ' ',
           's', 't', 'e', 'a', 'l']
print reverse_words(message)

File: InterviewCakeProblems2/reverse_words.rb
def reverse_words(message)
    message = reverse_str_in__place(message, 0, message.length - 1)

    i = 0
    startIdx = 0
    until i == message.length + 1
        char = message[i]
        if char == " " || i == message.length
            reverse_str_in__place(message, startIdx, i - 1)
            startIdx = i + 1
        end
        i += 1
    end
    message.join(" ")
end

def reverse_str_in__place(chars, start, endIdx)

    while start < endIdx
        temp = chars[start]
        chars[start] = chars[endIdx]
        chars[endIdx] = temp
        start += 1
        endIdx -= 1
    end
    chars
end

message = ['c', 'a', 'k', 'e', ' ',
    'p', 'o', 'u', 'n', 'd', ' ',
    's', 't', 'e', 'a', 'l'];
print reverse_words(message)